tfhe/src/integer/server_key/radix_parallel/modulus_switch_compression.rs

```diff
@@ -61,12 +61,96 @@ impl ServerKey {
         &self,
         compressed_ct: &CompressedModulusSwitchedSignedRadixCiphertext,
     ) -> SignedRadixCiphertext {
-        let blocks = compressed_ct
-            .blocks
-            .par_iter()
-            .map(|a| self.key.decompress(a))
-            .collect();
+        let blocks = uncompress_list_by_two(
+            self,
+            &compressed_ct.paired_blocks,
+            compressed_ct.last_block.as_ref(),
+        );
 
         BaseSignedRadixCiphertext { blocks }
     }
 }
+
+fn compress_list_by_two(
+    sk: &ServerKey,
+    blocks: &[Ciphertext],
+) -> (
+    Vec<CompressedModulusSwitchedCiphertext>,
+    Option<CompressedModulusSwitchedCiphertext>,
+) {
+    assert!(
+        sk.message_modulus().0 <= sk.carry_modulus().0,
+        "Compression does not support message_modulus > carry_modulus"
+    );
+
+    let paired_blocks;
+
+    let last_block;
+
+    let len = blocks.len();
+
+    if len % 2 == 1 {
+        paired_blocks = &blocks[..2 * (len / 2)];
+
+        last_block = Some(blocks.last().unwrap());
+    } else {
+        paired_blocks = blocks;
+        last_block = None;
+    }
+
+    let paired_blocks = paired_blocks
+        .par_iter()
+        .chunks(2)
+        .map(|pair| {
+            let mut packed = pair[0].clone();
+
+            if pair.len() == 2 {
+                let scaled = sk
+                    .key
+                    .unchecked_scalar_mul(pair[1], sk.message_modulus().0 as u8);
+
+                sk.key.unchecked_add_assign(&mut packed, &scaled);
+            }
+
+            sk.key.switch_modulus_and_compress(&packed)
+        })
+        .collect();
+
+    let last_block = last_block.map(|a| sk.key.switch_modulus_and_compress(a));
+
+    (paired_blocks, last_block)
+}
+
+fn uncompress_list_by_two(
+    sk: &ServerKey,
+    pairs: &[CompressedModulusSwitchedCiphertext],
+    last: Option<&CompressedModulusSwitchedCiphertext>,
+) -> Vec<Ciphertext> {
+    let message_extract = sk
+        .key
+        .generate_msg_lookup_table(|x| x, sk.message_modulus());
```

Comment:
I think we should retire generate_msg_lookup_table it is nearly never used and hides what the look up does, please use the plain generate_lookup_table here, as it indicates clearly we are extracting the message part from x

Reply:
Done